scripts/dev/gen_stub.php

```diff
@@ -0,0 +1,280 @@
+    foreach ($stmts as $stmt) {
+        if ($stmt instanceof Stmt\Nop) {
+            continue;
+        }
+
+        if ($stmt instanceof Stmt\Function_) {
+            $funcInfos[] = parseFunctionLike($stmt->name->toString(), $stmt);
+            continue;
+        }
+
+        if ($stmt instanceof Stmt\ClassLike) {
+            $className = $stmt->name->toString();
+            foreach ($stmt->stmts as $classStmt) {
+                if (!$classStmt instanceof Stmt\ClassMethod) {
+                    throw new Exception("Not implemented {$classStmt->getType()}");
+                }
+
+                $funcInfos[] = parseFunctionLike(
+                    $className . '_' . $classStmt->name->toString(), $classStmt);
+            }
+            continue;
+        }
+
+        // TODO Collect #if?
+        throw new Exception("Unexpected node {$stmt->getType()}");
+    }
+
+    return $funcInfos;
+}
+
+function funcInfoToCode(FuncInfo $funcInfo): string {
+    $code = '';
+    if ($funcInfo->return->type) {
+        $returnType = $funcInfo->return->type;
+        if ($returnType->isBuiltin) {
+            $code .= sprintf(
+                "ZEND_BEGIN_ARG_WITH_RETURN_TYPE_INFO_EX(arginfo_%s, %d, %d, %s, %d)\n",
+                $funcInfo->name, $funcInfo->return->byRef, $funcInfo->numRequiredArgs,
+                $returnType->toTypeCode(), $returnType->isNullable
+            );
+        } else {
+            throw new Exception("Not implemented");
+        }
+    } else {
+        $code .= sprintf(
+            "ZEND_BEGIN_ARG_INFO_EX(arginfo_%s, 0, %d, %d)\n",
+            $funcInfo->name, $funcInfo->return->byRef, $funcInfo->numRequiredArgs
+        );
+    }
+
+    foreach ($funcInfo->args as $argInfo) {
+        if ($argInfo->isVariadic) {
+            throw new Exception("Not implemented");
+        }
+        if ($argInfo->type) {
+            if ($argInfo->type->isBuiltin) {
+                $code .= sprintf(
+                    "\tZEND_ARG_TYPE_INFO(%d, %s, %s, %d)\n",
+                    $argInfo->byRef, $argInfo->name,
+                    $argInfo->type->toTypeCode(), $argInfo->type->isNullable
+                );
+            } else {
+                $code .= sprintf(
+                    "\tZEND_ARG_OBJ_INFO(%d, %s, %s, %d)\n",
+                    $argInfo->byRef, $argInfo->name,
+                    $argInfo->type->name, $argInfo->type->isNullable
+                );
+            }
+        } else {
+            $code .= sprintf("\tZEND_ARG_INFO(%d, %s)\n", $argInfo->byRef, $argInfo->name);
+        }
+    }
+
+    $code .= "ZEND_END_ARG_INFO()";
+    return $code;
+}
+
+/** @param FuncInfo[] $funcInfos */
+function generateArginfoCode(array $funcInfos): string {
+    $code = "/* This is a generated file, edit the .stub.php file instead. */\n\n";
+    return $code . implode("\n\n", array_map('funcInfoToCode', $funcInfos));
+}
+
+function initPhpParser() {
+    $version = "4.2.2";
+    if (!is_dir(__DIR__ . "/PHP-Parser")) {
```

Comment:
shouldn't you have a way to check whether the existing installed version is at the expected version (to automatically re-download it if the version expected by the script gets updated) ?

Reply:
the simplest solution: use versioned dir (`PHP-Parser-4.2.2`).

Reply:
or `PHP-Parser/4.2.2` to keep a single top-level folder than can be ignored in git (even though we could also ignore a wildcard)

Reply:
I've moved to using `PHP-Parsre-4.2.2`.

Reply:
I've moved this to use `PHP-Parser-4.2.2`.